File: galaxyos/kernel/dsl_bridge.py
"""
DSLBridge — TokUI DSL ↔ EUI-NEO DSL 双向转换引擎

将 GalaxyOS 的 TokUI DSL 转换为 EUI-NEO 原生渲染可理解的 DSL，
支持声明式映射表、动态组件注册、转换置信度评估和超时检测。
"""

from __future__ import annotations

import json
import logging
import os
import re
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

logger = logging.getLogger(__name__)

DEFAULT_MAPPINGS = {
    "card": {"eui_type": "card", "attrs": {"padding": "8", "radius": "4"}, "confidence": 1.0},
    "p": {"eui_type": "text", "attrs": {}, "confidence": 1.0},
    "btn": {"eui_type": "button", "attrs": {}, "confidence": 1.0},
    "table": {"eui_type": "dataTable", "attrs": {}, "confidence": 0.9},
    "chart": {"eui_type": "lineChart", "attrs": {}, "confidence": 0.8},
    "img": {"eui_type": "image", "attrs": {}, "confidence": 1.0},
    "input": {"eui_type": "input", "attrs": {}, "confidence": 1.0},
    "list": {"eui_type": "scrollView", "attrs": {}, "confidence": 0.8},
    "vlist": {"eui_type": "virtualList", "attrs": {}, "confidence": 0.9},
    "menu": {"eui_type": "contextMenu", "attrs": {}, "confidence": 0.9},
    "tabs": {"eui_type": "tabs", "attrs": {}, "confidence": 1.0},
    "dialog": {"eui_type": "dialog", "attrs": {}, "confidence": 1.0},
    "progress": {"eui_type": "progress", "attrs": {}, "confidence": 1.0},
    "markdown": {"eui_type": "markdown", "attrs": {}, "confidence": 1.0},
    "sidebar": {"eui_type": "sidebar", "attrs": {}, "confidence": 1.0},
    "slider": {"eui_type": "slider", "attrs": {}, "confidence": 1.0},
    "switch": {"eui_type": "switch", "attrs": {}, "confidence": 1.0},
    "checkbox": {"eui_type": "checkbox", "attrs": {}, "confidence": 1.0},
    "radio": {"eui_type": "radio", "attrs": {}, "confidence": 1.0},
    "dropdown": {"eui_type": "dropdown", "attrs": {}, "confidence": 1.0},
    "tooltip": {"eui_type": "tooltip", "attrs": {}, "confidence": 1.0},
    "toast": {"eui_type": "toast", "attrs": {}, "confidence": 1.0},
    "carousel": {"eui_type": "carousel", "attrs": {}, "confidence": 1.0},
    "stepper": {"eui_type": "stepper", "attrs": {}, "confidence": 1.0},
    "segmented": {"eui_type": "segmented", "attrs": {}, "confidence": 1.0},
    "date-picker": {"eui_type": "datePicker", "attrs": {}, "confidence": 1.0},
    "time-picker": {"eui_type": "timePicker", "attrs": {}, "confidence": 1.0},
    "color-picker": {"eui_type": "colorPicker", "attrs": {}, "confidence": 1.0},
    "bar-chart": {"eui_type": "barChart", "attrs": {}, "confidence": 1.0},
    "pie-chart": {"eui_type": "pieChart", "attrs": {}, "confidence": 1.0},
    "memory-panel": {"eui_type": "panel", "attrs": {"panel_type": "memory"}, "confidence": 1.0, "builder": "ProgressBuilder + InputBuilder + SegmentedBuilder"},
    "rccam-progress": {"eui_type": "panel", "attrs": {"panel_type": "rccam"}, "confidence": 1.0, "builder": "ProgressBuilder + ButtonBuilder + SliderBuilder + DropdownBuilder"},
    "dag-tree": {"eui_type": "panel", "attrs": {"panel_type": "dag_tree"}, "confidence": 1.0, "builder": "TextBuilder + ButtonBuilder (recursive)"},
    "memory-search": {"eui_type": "panel", "attrs": {"panel_type": "search"}, "confidence": 1.0, "builder": "InputBuilder + SegmentedBuilder + ScrollViewBuilder"},
    "rccam-control": {"eui_type": "panel", "attrs": {"panel_type": "rccam_control"}, "confidence": 1.0, "builder": "ButtonBuilder + SliderBuilder + DropdownBuilder"},
    "dag-node-expand": {"eui_type": "panel", "attrs": {"panel_type": "dag_expand"}, "confidence": 1.0, "builder": "TextBuilder + ButtonBuilder"},
    "cognitive-panel": {"eui_type": "panel", "attrs": {"panel_type": "cognitive"}, "confidence": 1.0, "builder": "SidebarBuilder + TabsBuilder"},
    "chat-renderer": {"eui_type": "panel", "attrs": {"panel_type": "chat"}, "confidence": 1.0, "builder": "ScrollViewBuilder + TextBuilder"},
    "message-renderer": {"eui_type": "panel", "attrs": {"panel_type": "message"}, "confidence": 1.0, "builder": "TextBuilder + MarkdownBuilder"},
}

CONVERSION_TIMEOUT_MS = 5.0


@dataclass
class DSLBridgeResult:
    output_dsl: str
    mapping_confidence: float
    unsupported_components: List[str]
    dropped_attrs: List[str]
    conversion_time_ms: float


class DSLBridge:
    def __init__(self, config_path: str = ""):
        self._mappings: Dict[str, Dict[str, Any]] = dict(DEFAULT_MAPPINGS)
        self._config_path = config_path or os.path.join(
            os.environ.get("GALAXYOS_HOME", os.path.join(os.path.dirname(__file__), "..", "..")),
            "config", "dsl_bridge_mappings.json"
        )
        self._load_mappings()

    def _load_mappings(self) -> None:
        if os.path.isfile(self._config_path):
            try:
                with open(self._config_path, "r", encoding="utf-8") as f:
                    custom = json.load(f)
                for k, v in custom.get("mappings", {}).items():
                    self._mappings[k] = v
                logger.info(f"Loaded {len(custom.get('mappings', {}))} custom DSL mappings from {self._config_path}")
            except Exception as e:
                logger.warning(f"Failed to load DSL mappings: {e}")

    def _save_mappings(self) -> None:
        try:
            os.makedirs(os.path.dirname(self._config_path), exist_ok=True)
            with open(self._config_path, "w", encoding="utf-8") as f:
                json.dump({"mappings": self._mappings}, f, indent=2, ensure_ascii=False)
        except Exception as e:
            logger.warning(f"Failed to save DSL mappings: {e}")

    def tokui_to_eui(self, tokui_dsl: str) -> DSLBridgeResult:
        start = time.time()
        unsupported = []
        dropped = []
        total_confidence = 0.0
        mapped_count = 0

        tokens = self._tokenize_tokui(tokui_dsl)
        eui_components = []

        for token in tokens:
            comp_type = token.get("type", "")
            mapping = self._mappings.get(comp_type)

            if not mapping:
                unsupported.append(comp_type)
                eui_components.append(self._fallback_component(token))
                continue

            eui_type = mapping.get("eui_type", "Unknown")
            confidence = mapping.get("confidence", 0.5)
            base_attrs = dict(mapping.get("attrs", {}))

            merged_attrs = {**base_attrs, **token.get("attrs", {})}
            for attr_name in list(merged_attrs.keys()):
                if attr_name.startswith("_"):
                    dropped.append(f"{comp_type}.{attr_name}")
                    del merged_attrs[attr_name]

            eui_components.append({
                "type": eui_type,
                "attrs": merged_attrs,
                "children": token.get("children", []),
                "content": token.get("content", ""),
            })

            total_confidence += confidence
            mapped_count += 1

        avg_confidence = total_confidence / max(mapped_count, 1)
        elapsed_ms = (time.time() - start) * 1000

        if elapsed_ms > CONVERSION_TIMEOUT_MS:
            logger.warning(f"DSL conversion took {elapsed_ms:.1f}ms (timeout: {CONVERSION_TIMEOUT_MS}ms)")

        output = json.dumps({"components": eui_components}, ensure_ascii=False)

        return DSLBridgeResult(
            output_dsl=output,
            mapping_confidence=round(avg_confidence, 3),
            unsupported_components=unsupported,
            dropped_attrs=dropped,
            conversion_time_ms=round(elapsed_ms, 2),
        )

    def eui_to_tokui(self, eui_dsl: str) -> DSLBridgeResult:
        start = time.time()
        unsupported = []
        dropped = []
        total_confidence = 0.0
        mapped_count = 0

        reverse_map = {}
        for tokui_type, mapping in self._mappings.items():
            eui_type = mapping.get("eui_type", "")
            if eui_type:
                reverse_map[eui_type] = {"tokui_type": tokui_type, "confidence": mapping.get("confidence", 0.5)}

        try:
            eui_data = json.loads(eui_dsl) if isinstance(eui_dsl, str) else eui_dsl
        except json.JSONDecodeError:
            return DSLBridgeResult(
                output_dsl="", mapping_confidence=0, unsupported_components=["invalid_json"],
                dropped_attrs=[], conversion_time_ms=0,
            )

        tokui_components = []
        for comp in eui_data.get("components", []):
            eui_type = comp.get("type", "")
            rev = reverse_map.get(eui_type)

            if not rev:
                unsupported.append(eui_type)
                tokui_components.append(f"[{eui_type}]")
                continue

            tokui_type = rev["tokui_type"]
            confidence = rev["confidence"]
            attrs = comp.get("attrs", {})
            content = comp.get("content", "")

            attr_str = " ".join(f'{k}:{v}' for k, v in attrs.items() if not k.startswith("_"))
            tokui_components.append(f"[{tokui_type}{' ' + attr_str if attr_str else ''}]{content}[/{tokui_type}]")

            total_confidence += confidence
            mapped_count += 1

        avg_confidence = total_confidence / max(mapped_count, 1)
        elapsed_ms = (time.time() - start) * 1000

        return DSLBridgeResult(
            output_dsl="".join(tokui_components),
            mapping_confidence=round(avg_confidence, 3),
            unsupported_components=unsupported,
            dropped_attrs=dropped,
            conversion_time_ms=round(elapsed_ms, 2),
        )

    def register_component_mapping(
        self, tokui_type: str, eui_type: str, attr_mapping: Optional[Dict[str, Any]] = None, confidence: float = 0.5
    ) -> None:
        self._mappings[tokui_type] = {
            "eui_type": eui_type,
            "attrs": attr_mapping or {},
            "confidence": confidence,
        }
        self._save_mappings()
        logger.info(f"Registered DSL mapping: {tokui_type} → {eui_type} (confidence={confidence})")

    def get_unsupported_components(self, dsl: str) -> List[str]:
        tokens = self._tokenize_tokui(dsl)
        return [t.get("type", "") for t in tokens if t.get("type", "") not in self._mappings]

    def get_mapping_stats(self) -> Dict[str, Any]:
        cognitive = [k for k in self._mappings if k in DEFAULT_MAPPINGS and DEFAULT_MAPPINGS[k].get("confidence", 0) >= 0.6]
        basic = [k for k in self._mappings if k in DEFAULT_MAPPINGS and k not in cognitive]
        custom = [k for k in self._mappings if k not in DEFAULT_MAPPINGS]
        return {
            "total_mappings": len(self._mappings),
            "cognitive_components": len(cognitive),
            "basic_components": len(basic),
            "custom_mappings": len(custom),
            "cognitive_list": cognitive,
            "basic_list": basic,
            "custom_list": custom,
        }

    def _tokenize_tokui(self, dsl: str) -> List[Dict[str, Any]]:
        tokens = []
        pattern = r'\[(\S+?)(?:\s+([^\]]*))?\](.*?)\[/\1\]'
        for match in re.finditer(pattern, dsl, re.DOTALL):
            comp_type = match.group(1)
            attr_str = match.group(2) or ""
            content = match.group(3).strip()

            attrs = {}
            for attr_pair in re.findall(r'(\w+):([^\s]+)', attr_str):
                attrs[attr_pair[0]] = attr_pair[1]

            tokens.append({"type": comp_type, "attrs": attrs, "content": content, "children": []})

        if not tokens:
            tokens.append({"type": "p", "attrs": {}, "content": dsl.strip(), "children": []})

        return tokens

    def _fallback_component(self, token: Dict[str, Any]) -> Dict[str, Any]:
        return {
            "type": "Text",
            "attrs": {"fallback": True, "original_type": token.get("type", "unknown")},
            "content": token.get("content", ""),
            "children": [],
        }

File: desktop-tauri/src/dsl_bridge.rs
use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::time::Instant;

const CONVERSION_TIMEOUT_MS: f64 = 5.0;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DslBridgeResult {
    pub output_dsl: String,
    pub mapping_confidence: f64,
    pub unsupported_components: Vec<String>,
    pub dropped_attrs: Vec<String>,
    pub conversion_time_ms: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct MappingEntry {
    eui_type: String,
    #[serde(default)]
    attrs: HashMap<String, Value>,
    #[serde(default = "default_confidence")]
    confidence: f64,
    #[serde(default)]
    builder: Option<String>,
}

fn default_confidence() -> f64 {
    0.5
}

fn default_mappings() -> HashMap<&'static str, MappingEntry> {
    let mut m = HashMap::new();
    let basic = [
        ("card", "card", 1.0),
        ("p", "text", 1.0),
        ("btn", "button", 1.0),
        ("table", "dataTable", 0.9),
        ("chart", "lineChart", 0.8),
        ("img", "image", 1.0),
        ("input", "input", 1.0),
        ("list", "scrollView", 0.8),
        ("vlist", "virtualList", 0.9),
        ("menu", "contextMenu", 0.9),
        ("tabs", "tabs", 1.0),
        ("dialog", "dialog", 1.0),
        ("progress", "progress", 1.0),
        ("markdown", "markdown", 1.0),
        ("sidebar", "sidebar", 1.0),
        ("slider", "slider", 1.0),
        ("switch", "switch", 1.0),
        ("checkbox", "checkbox", 1.0),
        ("radio", "radio", 1.0),
        ("dropdown", "dropdown", 1.0),
        ("tooltip", "tooltip", 1.0),
        ("toast", "toast", 1.0),
        ("carousel", "carousel", 1.0),
        ("stepper", "stepper", 1.0),
        ("segmented", "segmented", 1.0),
        ("date-picker", "datePicker", 1.0),
        ("time-picker", "timePicker", 1.0),
        ("color-picker", "colorPicker", 1.0),
        ("bar-chart", "barChart", 1.0),
        ("pie-chart", "pieChart", 1.0),
    ];
    for (tokui, eui, conf) in basic {
        m.insert(tokui, MappingEntry {
            eui_type: eui.to_string(),
            attrs: HashMap::new(),
            confidence: conf,
            builder: None,
        });
    }

    let cognitive = [
        ("memory-panel", "panel", 1.0, "ProgressBuilder + InputBuilder + SegmentedBuilder"),
        ("rccam-progress", "panel", 1.0, "ProgressBuilder + ButtonBuilder + SliderBuilder + DropdownBuilder"),
        ("dag-tree", "panel", 1.0, "TextBuilder + ButtonBuilder (recursive)"),
        ("memory-search", "panel", 1.0, "InputBuilder + SegmentedBuilder + ScrollViewBuilder"),
        ("rccam-control", "panel", 1.0, "ButtonBuilder + SliderBuilder + DropdownBuilder"),
        ("dag-node-expand", "panel", 1.0, "TextBuilder + ButtonBuilder"),
        ("cognitive-panel", "panel", 1.0, "SidebarBuilder + TabsBuilder"),
        ("chat-renderer", "panel", 1.0, "ScrollViewBuilder + TextBuilder"),
        ("message-renderer", "panel", 1.0, "TextBuilder + MarkdownBuilder"),
    ];
    for (tokui, eui, conf, builder) in cognitive {
        let mut attrs = HashMap::new();
        attrs.insert("panel_type".to_string(), Value::String(tokui.replace("-panel", "").replace("-progress", "").replace("-tree", "").replace("-search", "").replace("-control", "").replace("-expand", "").replace("-renderer", "").to_string()));
        m.insert(tokui, MappingEntry {
            eui_type: eui.to_string(),
            attrs,
            confidence: conf,
            builder: Some(builder.to_string()),
        });
    }

    m
}

#[derive(Debug, Clone)]
struct TokenizedComponent {
    comp_type: String,
    attrs: HashMap<String, String>,
    content: String,
}

pub struct DslBridge {
    mappings: HashMap<String, MappingEntry>,
}

impl DslBridge {
    pub fn new() -> Self {
        let mut mappings = HashMap::new();
        for (k, v) in default_mappings() {
            mappings.insert(k.to_string(), v);
        }
        Self { mappings }
    }

    pub fn tokui_to_eui(&self, tokui_dsl: &str) -> DslBridgeResult {
        let start = Instant::now();
        let mut unsupported = Vec::new();
        let mut dropped = Vec::new();
        let mut total_confidence = 0.0;
        let mut mapped_count = 0;

        let tokens = self.tokenize_tokui(tokui_dsl);
        let mut eui_components = Vec::new();

        for token in &tokens {
            let mapping = self.mappings.get(&token.comp_type);

            if mapping.is_none() {
                unsupported.push(token.comp_type.clone());
                eui_components.push(self.fallback_component(token));
                continue;
            }

            let m = mapping.unwrap();
            let mut merged_attrs = m.attrs.clone();
            for (k, v) in &token.attrs {
                merged_attrs.insert(k.clone(), Value::String(v.clone()));
            }

            let mut dropped_keys = Vec::new();
            let keys: Vec<String> = merged_attrs.keys().cloned().collect();
            for k in keys {
                if k.starts_with('_') {
                    dropped.push(format!("{}.{}", token.comp_type, k));
                    dropped_keys.push(k);
                }
            }
            for k in dropped_keys {
                merged_attrs.remove(&k);
            }

            eui_components.push(serde_json::json!({
                "type": m.eui_type,
                "attrs": merged_attrs,
                "children": [],
                "content": token.content,
            }));

            total_confidence += m.confidence;
            mapped_count += 1;
        }

        let avg_confidence = total_confidence / mapped_count.max(1) as f64;
        let elapsed_ms = start.elapsed().as_secs_f64() * 1000.0;

        if elapsed_ms > CONVERSION_TIMEOUT_MS {
            log::warn!("DSL conversion took {:.1}ms (timeout: {}ms)", elapsed_ms, CONVERSION_TIMEOUT_MS);
        }

        let output = serde_json::json!({"components": eui_components}).to_string();

        DslBridgeResult {
            output_dsl: output,
            mapping_confidence: (avg_confidence * 1000.0).round() / 1000.0,
            unsupported_components: unsupported,
            dropped_attrs: dropped,
            conversion_time_ms: (elapsed_ms * 100.0).round() / 100.0,
        }
    }

    pub fn eui_to_tokui(&self, eui_dsl: &str) -> DslBridgeResult {
        let start = Instant::now();
        let mut unsupported = Vec::new();
        let mut total_confidence = 0.0;
        let mut mapped_count = 0;

        let mut reverse_map: HashMap<&str, (&str, f64)> = HashMap::new();
        for (tokui_type, mapping) in &self.mappings {
            reverse_map.insert(&mapping.eui_type, (tokui_type.as_str(), mapping.confidence));
        }

        let eui_data: Value = match serde_json::from_str(eui_dsl) {
            Ok(v) => v,
            Err(_) => {
                return DslBridgeResult {
                    output_dsl: String::new(),
                    mapping_confidence: 0.0,
                    unsupported_components: vec!["invalid_json".into()],
                    dropped_attrs: Vec::new(),
                    conversion_time_ms: 0.0,
                }
            }
        };

        let mut tokui_components = Vec::new();

        if let Some(components) = eui_data.get("components").and_then(|c| c.as_array()) {
            for comp in components {
                let eui_type = comp.get("type").and_then(|t| t.as_str()).unwrap_or("");
                let rev = reverse_map.get(eui_type);

                if rev.is_none() {
                    unsupported.push(eui_type.to_string());
                    tokui_components.push(format!("[{}]", eui_type));
                    continue;
                }

                let (tokui_type, confidence) = rev.unwrap();
                let attrs = comp.get("attrs").and_then(|a| a.as_object());
                let content = comp.get("content").and_then(|c| c.as_str()).unwrap_or("");

                let attr_str = attrs.map(|a| {
                    let pairs: Vec<String> = a.iter()
                        .filter(|(k, _)| !k.starts_with('_'))
                        .map(|(k, v)| format!("{}:{}", k, v))
                        .collect();
                    pairs.join(" ")
                }).unwrap_or_default();

                if attr_str.is_empty() {
                    tokui_components.push(format!("[{}]{}[/{}/]", tokui_type, content, tokui_type));
                } else {
                    tokui_components.push(format!("[{} {}]{}[/{}/]", tokui_type, attr_str, content, tokui_type));
                }

                total_confidence += confidence;
                mapped_count += 1;
            }
        }

        let avg_confidence = total_confidence / mapped_count.max(1) as f64;
        let elapsed_ms = start.elapsed().as_secs_f64() * 1000.0;

        DslBridgeResult {
            output_dsl: tokui_components.join(""),
            mapping_confidence: (avg_confidence * 1000.0).round() / 1000.0,
            unsupported_components: unsupported,
            dropped_attrs: Vec::new(),
            conversion_time_ms: (elapsed_ms * 100.0).round() / 100.0,
        }
    }

    pub fn register_mapping(
        &mut self,
        tokui_type: &str,
        eui_type: &str,
        confidence: f64,
    ) {
        self.mappings.insert(tokui_type.to_string(), MappingEntry {
            eui_type: eui_type.to_string(),
            attrs: HashMap::new(),
            confidence,
            builder: None,
        });
        log::info!("Registered DSL mapping: {} -> {} (confidence={})", tokui_type, eui_type, confidence);
    }

    pub fn get_mapping_stats(&self) -> serde_json::Value {
        let total = self.mappings.len();
        let cognitive: Vec<&String> = self.mappings.keys()
            .filter(|k| self.mappings.get(*k).map_or(false, |m| m.confidence >= 0.6 && m.builder.is_some()))
            .collect();
        let basic: Vec<&String> = self.mappings.keys()
            .filter(|k| self.mappings.get(*k).map_or(false, |m| m.builder.is_none()))
            .collect();

        serde_json::json!({
            "total_mappings": total,
            "cognitive_components": cognitive.len(),
            "basic_components": basic.len(),
        })
    }

    fn tokenize_tokui(&self, dsl: &str) -> Vec<TokenizedComponent> {
        let mut tokens = Vec::new();
        let re = Regex::new(r"\[(\S+?)(?:\s+([^\]]*))?\](.*?)\[/\1\]").unwrap();

        for cap in re.captures_iter(dsl) {
            let comp_type = cap.get(1).map(|m| m.as_str().to_string()).unwrap_or_default();
            let attr_str = cap.get(2).map(|m| m.as_str()).unwrap_or("");
            let content = cap.get(3).map(|m| m.as_str().trim().to_string()).unwrap_or_default();

            let mut attrs = HashMap::new();
            let attr_re = Regex::new(r"(\w+):([^\s]+)").unwrap();
            for attr_cap in attr_re.captures_iter(attr_str) {
                if let (Some(k), Some(v)) = (attr_cap.get(1), attr_cap.get(2)) {
                    attrs.insert(k.as_str().to_string(), v.as_str().to_string());
                }
            }

            tokens.push(TokenizedComponent {
                comp_type,
                attrs,
                content,
            });
        }

        if tokens.is_empty() {
            tokens.push(TokenizedComponent {
                comp_type: "p".to_string(),
                attrs: HashMap::new(),
                content: dsl.trim().to_string(),
            });
        }

        tokens
    }

    fn fallback_component(&self, token: &TokenizedComponent) -> Value {
        serde_json::json!({
            "type": "Text",
            "attrs": {"fallback": true, "original_type": token.comp_type},
            "content": token.content,
            "children": [],
        })
    }
}
